File: csharp/src/AnQL.Functions/AnQLFunctionsVisitor.cs
using AnQL.Core;
using AnQL.Core.Grammar;
using AnQL.Core.Resolvers;

namespace AnQL.Functions;

public class AnQLFunctionsVisitor<T> : AnQLBaseVisitor<T, Func<T, bool>>
{
    private static readonly Func<T, bool> AlwaysTrue = _ => true;
    private static readonly Func<T, bool> AlwaysFalse = _ => false;
    
    public override Func<T, bool> SuccessQueryResult => AlwaysTrue;
    public override Func<T, bool> FailedQueryResult => AlwaysFalse;

    public AnQLFunctionsVisitor(ResolverMap<Func<T, bool>, T> resolverMap, AnQLParserOptions options) : base(resolverMap, options)
    {
    }
    
    public override Func<T, bool> VisitAnd(Func<T, bool> left, Func<T, bool> right) =>
        value => left(value) && right(value);

    public override Func<T, bool> VisitOr(Func<T, bool> left, Func<T, bool> right) =>
        value => left(value) || right(value);

    public override Func<T, bool> VisitNot(Func<T, bool> childExpression) =>
        value => !childExpression(value);
}

File: typescript/functions/src/AnQLFunctionsVisitor.ts
import { AnQLBaseVisitor } from '@anql/core/dist'
import { Predicate } from '@src/types'

export class AnQLFunctionsVisitor<TModel> extends AnQLBaseVisitor<TModel, Predicate<TModel>> {
    successQueryResult: Predicate<TModel> = _ => true
    failedQueryResult: Predicate<TModel> = _ => false
    public handleAnd(left: Predicate<TModel>, right: Predicate<TModel>): Predicate<TModel> {
        return arg => left(arg) && right(arg);
    }
    public handleOr(left: Predicate<TModel>, right: Predicate<TModel>): Predicate<TModel> {
        return arg => left(arg) || right(arg)
    }
    public handleNot(childExpression: Predicate<TModel>): Predicate<TModel> {
        return arg => !childExpression(arg)
    }
    public handleAnyEqual(...childExpressions: Predicate<TModel>[]): Predicate<TModel> {
        return childExpressions.reduce(this.handleOr);
    }
}
